File: apps/Camera-Vitals/iOS/CameraVitals/Signal/Detrend.swift
import Foundation

/// Removes a least-squares linear trend (baseline drift) from a signal.
/// Combined with the zero-phase bandpass downstream, this matches the
/// peak-frequency behaviour of the toolbox's smoothness-prior detrend.
enum Detrend {
    static func linear(_ x: [Float]) -> [Float] {
        let n = x.count
        guard n > 2 else { return x }
        let nF = Float(n)
        var sumT: Float = 0, sumT2: Float = 0, sumY: Float = 0, sumTY: Float = 0
        for i in 0..<n {
            let t = Float(i)
            sumT += t; sumT2 += t * t; sumY += x[i]; sumTY += t * x[i]
        }
        let denom = nF * sumT2 - sumT * sumT
        guard abs(denom) > 1e-9 else { return x }
        let b = (nF * sumTY - sumT * sumY) / denom
        let a = (sumY - b * sumT) / nF
        var out = [Float](repeating: 0, count: n)
        for i in 0..<n { out[i] = x[i] - (a + b * Float(i)) }
        return out
    }
}

File: apps/Camera-Vitals/Android/app/src/main/java/ai/zetic/demo/cameravitals/signal/Detrend.kt
package ai.zetic.demo.cameravitals.signal

import kotlin.math.abs

/** Removes a least-squares linear trend (baseline drift). Mirrors iOS Detrend.linear. */
object Detrend {
    fun linear(x: FloatArray): FloatArray {
        val n = x.size
        if (n <= 2) return x
        val nF = n.toFloat()
        var sumT = 0f; var sumT2 = 0f; var sumY = 0f; var sumTY = 0f
        for (i in 0 until n) {
            val t = i.toFloat()
            sumT += t; sumT2 += t * t; sumY += x[i]; sumTY += t * x[i]
        }
        val denom = nF * sumT2 - sumT * sumT
        if (abs(denom) < 1e-9f) return x
        val b = (nF * sumTY - sumT * sumY) / denom
        val a = (sumY - b * sumT) / nF
        return FloatArray(n) { i -> x[i] - (a + b * i) }
    }
}
